packages/RawButton/src/RawButton.js

```diff
@@ -46,18 +46,28 @@ const RawButton = React.forwardRef((props, ref) => {
   if (a11yText) moreProps["aria-label"] = a11yText;
 
   const rawButtonRef = React.useRef(null);
+  const [hasForcedFocus, setHasForcedFocus] = React.useState(false);
 
   React.useImperativeHandle(ref, () => ({
     focus: () => {
+      setHasForcedFocus(true);
       rawButtonRef.current.focus();
     },
   }));
 
   const handleClick = event => {
+    setHasForcedFocus(false);
```

Comment:
I don't think you need react state at all. you can simply use regular dom and `setAttribute`

like 
```js
// adding
$dom.setAttribute("data-has-forced-focus", true)
// removing
$dom.removeAttribute("data-has-forced-focus);
```

Reply:
Using state is consistent with the `<Button>` though:

https://github.com/acl-services/paprika/blob/master/packages/Button/src/Button.js#L118

And is arguably more readable? Maybe @jamiek-acl had other reasons for using state?

Reply:
??? but state it's not even needed here, what readability has to do with use proper javascript? just curious.

Reply:
I have changed it (and was also going to do so in the Button component), i am not completely clear on the pros versus cons of this approach of using a state change though. @nahumzs can you expand on that?

Reply:
I'm not _sure_ but if you do `$dom.setAttribute()` that will manipulate the DOM right now, vs setting state, which will leave it to React do decide when to manipulate the DOM.  I know manipulating the DOM is slow, so if we can leave it to React, won't it make the best (most efficient) decision about when it should really do it?

Reply:
`"data-has-forced-focus"` it's not a state its a `css class / attribute target` that you are adding to modify your element visually,  you don't need to add a state to add that attribute in your dom element you can do it with simple javascript. 

> DOM is slow is not slow 

You already have the ref to the dom element you want to add and remove the attribute doesn't see the perf issue in this such small interaction. And In the end, React also needs to update the Dom. 

But I'm not hard on this to me was just simpler comment in the spirit to write simpler code and avoid useless state variables.

Reply:
@nahumzs Ok, Just so i understand this for moving forward, are you saying that if a component has to have a class or any attribute that only changes the component visually _and_ there is a ref available you prefer we use `setAttribute` over a state variable?
I don't think we could use that approach for setting a class though.

Reply:
It's just for this case and it is not that I prefer a thing over the other one. :)

Reply:
thoughts:
> "data-has-forced-focus" it's not a state its a css class / attribute target that you are adding to modify your element visually

I think something like this absolutely is a state. "Modify your element visually" is not an indicator of whether something is state or not. In fact, _everything_ that governs what an element looks like on the screen is state in React _by default_ (if it's not a prop but you catch my drift). That's literally what state is for, to be able to change what's presented on the screen. You set the state => the component re-renders, potentially painting a different set of pixels.

Sidestepping this flow has to have justification. Because after you modify the DOM directly, React no longer knows what's in there (its vdom doesn't match the real one). And React docs even [say](https://reactjs.org/docs/integrating-with-other-libraries.html#integrating-with-jquery-chosen-plugin)
> It is important that if you modify the DOM outside of React flow, you must ensure React doesn’t have a reason to touch those DOM nodes.

I can think of just a few cases when that could be necessary:
- high-performance stuff
- integrating with third-party libs
- perhaps some weird animation shenanigans.

This component is literally a button. In my eyes, not using state here doesn't simplify, it complicates. Having or not having a ref to the element is inconsequential.

> if you do $dom.setAttribute() that will manipulate the DOM right now, vs setting state, which will leave it to React do decide when to manipulate the DOM. I know manipulating the DOM is slow

Yeah, like Nahum said React would basically do the same operation for you anyway. Also, I've seen people misunderstand the time scale of what it takes for a browser to do stuff, perhaps due to vague language in the docs and articles. I believe, unless it results in some complex CSS being applied, setting a single attribute on a DOM element is a matter of milliseconds. Unless you're doing this hundreds and thousands of times in a row, this is literally indistinguishable.

Reply:
Ok, Personally I vote we take the approach that is most consistent and in this case I feel using state is certainly a more consistent pattern that is easier for most devs to follow, so will go with that for now.

Reply:
I think you are misinterpreting my point or I didn't explain it correctly. And agree it doesn't matter if it's a ref or not. 

My point is that in this particular example there it's not moment when react need to know about the attribute, the OUTPUT of the component will not change with our without the attribute visually speaking.

On the markup the attribute will be added or removed. But visually the output will be always the same.

This visual representation is changing only because there is a class declare **outside of the component** that changes because of that attribute. 

To me, there is no point (apart from doing it in a "react way") to add a state.

> React doesn’t have a reason to touch those DOM nodes.

 ^ and react doesn't need to know about it, it's not gaining anything from knowing it. 

> This component is literally a button. In my eyes, not using state here doesn't simplify, it > complicates

In my eyes, otherwise, actually adding a state add complexity for something that can be solved with two lines of code. But I agree that this can be seen as complexity for some people since it's not the direct way of doing it, or the "react way". But that doesn't mean it is not completely valid and in my eyes even better. :) 

But I guess for the sake of "consistency" with other components we can use state even if add more code.